lib/libsys/lseek.2
.\" Copyright (c) 1980, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd July 13, 2020
.Dt LSEEK 2
.Os
.Sh NAME
.Nm lseek
.Nd reposition read/write file offset
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In unistd.h
.Ft off_t
.Fn lseek "int fildes" "off_t offset" "int whence"
.Sh DESCRIPTION
The
.Fn lseek
system call repositions the offset of the file descriptor
.Fa fildes
to the
argument
.Fa offset
according to the directive
.Fa whence .
The argument
.Fa fildes
must be an open
file descriptor.
The
.Fn lseek
system call
repositions the file position pointer associated with the file
descriptor
.Fa fildes
as follows:
.Bl -item -offset indent
.It
If
.Fa whence
is
.Dv SEEK_SET ,
the offset is set to
.Fa offset
bytes.
.It
If
.Fa whence
is
.Dv SEEK_CUR ,
the offset is set to its current location plus
.Fa offset
bytes.
.It
If
.Fa whence
is
.Dv SEEK_END ,
the offset is set to the size of the
file plus
.Fa offset
bytes.
.It
If
.Fa whence
is
.Dv SEEK_HOLE ,
the offset is set to the start of the next hole greater than or equal
to the supplied
.Fa offset .
The definition of a hole is provided below.
.It
If
.Fa whence
is
.Dv SEEK_DATA ,
the offset is set to the start of the next non-hole file region greater
than or equal to the supplied
.Fa offset .
.El
.Pp
The
.Fn lseek
system call allows the file offset to be set beyond the end
of the existing end-of-file of the file.
If data is later written
at this point, subsequent reads of the data in the gap return
bytes of zeros (until data is actually written into the gap).
However, the
.Fn lseek
system call does not, by itself, extend the size of a file.
.Pp
A
.Qq hole
is defined as a contiguous range of bytes in a file, all having the value of
zero, but not all zeros in a file are guaranteed to be represented as holes
returned with
.Dv SEEK_HOLE .
File systems are allowed to expose ranges of zeros with
.Dv SEEK_HOLE ,
but not required to.
Applications can use
.Dv SEEK_HOLE
to optimise their behavior for ranges of zeros, but must not depend on it to
find all such ranges in a file.
Each file is presented as having a zero-size virtual hole at the very
end of the file.
The existence of a hole at the end of every data region allows for easy
programming and also provides compatibility to the original implementation
in Solaris.
It also causes the current file size (i.e., end-of-file offset) to be returned
to indicate that there are no more holes past the supplied
.Fa offset .
Applications should use
.Fn fpathconf _PC_MIN_HOLE_SIZE
or
.Fn pathconf _PC_MIN_HOLE_SIZE
to determine if a file system supports
.Dv SEEK_HOLE .
See
.Xr pathconf 2 .
.Pp
For file systems that do not supply information about holes, the file will be
represented as one entire data region.
.Sh RETURN VALUES
Upon successful completion,
.Fn lseek
returns the resulting offset location as measured in bytes from the
beginning of the file.
Otherwise,
a value of -1 is returned and
.Va errno
is set to indicate
the error.
.Sh ERRORS
The
.Fn lseek
system call
will fail and the file position pointer will remain unchanged if:
.Bl -tag -width Er
.It Bq Er EBADF
The
.Fa fildes
argument
is not an open file descriptor.
.It Bq Er EINVAL
The
.Fa whence
argument
is not a proper value
or the resulting file offset would
be negative for a non-character special file.
.It Bq Er ENXIO
For
.Dv SEEK_DATA ,
there are no more data regions past the supplied offset.
Due to existence of the hole at the end of the file, for
.Dv SEEK_HOLE
this error is only returned when the
.Fa offset
already points to the end-of-file position.
.It Bq Er EOVERFLOW
The resulting file offset would be a value which cannot be represented
correctly in an object of type
.Fa off_t .
.It Bq Er ESPIPE
The
.Fa fildes
argument
is associated with a pipe, socket, or FIFO.
.El
.Sh SEE ALSO
.Xr dup 2 ,
.Xr open 2 ,
.Xr pathconf 2
.Sh STANDARDS
The
.Fn lseek
system call is expected to conform to
.St -p1003.1-2008 .
.Pp
The
.Dv SEEK_HOLE
and
.Dv SEEK_DATA
directives, along with the
.Er ENXIO
error, are expected to conform to
.St -p1003.1-2024 .
.Sh HISTORY
The
.Fn lseek
function appeared in
.At v7 .
.Sh BUGS
If the
.Fn lseek
system call is operating on a device which is incapable of seeking,
it will request the seek operation and return successfully,
even though no seek was performed.
Because the
.Ar offset
argument will be stored unconditionally in the file descriptor of that device,
there is no way to confirm if the seek operation succeeded or not
(e.g. using the
.Fn ftell
function).
Device types which are known to be incapable of seeking include
tape drives.
.Pp
The
.Fn lseek
system call will not detect whether media are present in changeable
media devices such as DVD or Blu-ray devices.
A requested seek operation will therefore return sucessfully when no
medium is present.
.Pp
This document's use of
.Fa whence
is incorrect English, but is maintained for historical reasons.


NEWS
May 20, 2025: libarchive 3.8.0 released

Mar 30, 2025: libarchive 3.7.9 released

Mar 20, 2025: libarchive 3.7.8 released

Oct 13, 2024: libarchive 3.7.7 released

Sep 23, 2024: libarchive 3.7.6 released

Sep 13, 2024: libarchive 3.7.5 released

Apr 26, 2024: libarchive 3.7.4 released

Apr 08, 2024: libarchive 3.7.3 released

Sep 12, 2023: libarchive 3.7.2 released

Jul 29, 2023: libarchive 3.7.1 released

Jul 18, 2023: libarchive 3.7.0 released

Jul 14, 2023: bsdunzip port from FreeBSD

Dec 07, 2022: libarchive 3.6.2 released

Apr 08, 2022: libarchive 3.6.1 released

Feb 09, 2022: libarchive 3.6.0 released

Feb 08, 2022: libarchive 3.5.3 released

Aug 22, 2021: libarchive 3.5.2 released

Dec 26, 2020: libarchive 3.5.1 released

Dec 01, 2020: libarchive 3.5.0 released

Oct 14, 2020: Support for system extended attributes

May 20, 2020: libarchive 3.4.3 released

Apr 30, 2020: Support for pzstd compressed files

Apr 16, 2020: Support for RHT.security.selinux tar extended attribute

Feb 11, 2020: libarchive 3.4.2 released

Jan 23, 2020: Important fixes for writing XAR archives

Jan 20, 2020: New tar option: --safe-writes (atomical file extraction)

Jan 03, 2020: Support mbed TLS (PolarSSL) as optional crypto provider

Dec 30, 2019: libarchive 3.4.1 released

Dec 11, 2019: New pax write option "xattrhdr"

Nov 17, 2019: Unicode filename support for reading lha/lzh archives

Jun 11, 2019: libarchive 3.4.0 released

May 18, 2019: Fixes for reading Android APK and JAR archives

Apr 16, 2019: Support for non-recursive list and extract

Apr 14, 2019: New tar option: --exclude-vcs

Mar 27, 2019: Support for file and directory symlinks on Windows

Mar 12, 2019: Important fixes for storing file attributes and flags

Jan 20, 2019: Support for xz, lzma, ppmd8 and bzip2 decompression in ZIP files

Oct 06, 2018: RAR 5.0 reader

Sep 03, 2018: libarchive 3.3.3 released

Jul 19, 2018: Avoid super-linear slowdown on malformed mtree files

Jan 27, 2018: Many fixes for building with Visual Studio

Oct 19, 2017: NO_OVERWRITE doesn't change existing directory attributes

Aug 12, 2017: New support for Zstandard read and write filters

Jul 09, 2017: libarchive 3.3.2 released

Mar 16, 2017: NFSv4 ACL support for Linux (librichacl)

Feb 26, 2017: libarchive 3.3.1 released
    Security & Feature release

Feb 19, 2017: libarchive 3.3.0 released
    Security & Feature release

Jan 29, 2017: Limited NFSv4 ACL support for Mac OS (Darwin)

Jan 10, 2017: POSIX.1e and NFSv4 ACL support for Solaris and derivates

Dec 27, 2016: NFSv4 ACL read and write support for pax
    Deprecated functions: archive_entry_acl_text(), archive_entry_acl_text_w()

Nov, 2016: libarchive is now being tested by the OSS-Fuzz project

Oct 26, 2016: Remove liblzmadec support

Oct 23, 2016: libarchive 3.2.2 released
    Security release

Jun 20, 2016: libarchive 3.2.1 released
    This fixes a handful of security and other critical issues with 3.2.0

May 01, 2016: libarchive 3.2.0 released

Apr 09, 2016: libarchive 3.1.901a released
    Another test release in preparation for 3.2.0

Feb 13, 2016: libarchive 3.1.900a released
    This is a test release in preparation for 3.2.0

Oct 21, 2015: Preliminary port to OSF

Apr 11, 2015: libarchive's issue tracker is now hosted at GitHub.
    https://github.com/libarchive/libarchive/issues

Early 2015:  Many fixes to crash and overflow bugs thanks to Hanno Boeck

Oct 13, 2014: Zip encryption and decryption support

Aug 13, 2014: Add support for lz4 compression.

Jun 10, 2014: Add warc format support

May  3, 2014: Add experimental Zip streaming extension

Apr  6, 2014: Add bsdcat command-line tool

Jan 12, 2014: Add Zip64 support

Dec  1, 2013: Rewrite Zip write logic

Jul  1, 2013: Add ability to detect encrypted entries for many formats
     (This does not add the ability to *decrypt* those entries, however)

Feb 23, 2013: "raw" write support added

Feb 09, 2013: libarchive 3.1.2 released

Jan 28, 2013: libarchive's new website moved to http://www.libarchive.org.

Jan 13, 2013: libarchive 3.1.1 released

Jan 13, 2013: libarchive 3.1.0 released

Dec 07, 2012: Implement functions to manually set the format and filters used.
    
Nov 11, 2012: Add support for __MACOSX directory in Zip archives, which resource
    forks are stored in.

Oct 20, 2012: Add support for writing v7 tar format.

Oct 09, 2012: Add support for grzip compression.

Oct 07, 2012: Introduce b64encode filter.    
Oct 07, 2012: Introduce uuencode filter.

Oct 06, 2012: Add support for lzop.
    
Sep 27, 2012: Implement function used to seek within data blocks.
    (Currently only supported for uncompressed RAR archives).
    
Apr 22, 2012: Add basic archive read and write filter support for lrzip.

Mar 27, 2012: libarchive 3.0.4 released

Feb 05, 2012: libarchive development now hosted at GitHub.
    http://libarchive.github.com/
Feb 05, 2012: libarchive's issue tracker remains at Google Code.
    http://code.google.com/p/libarchive/issues/list
Feb 05, 2012: libarchive's mailing lists remain at Google Groups.

Dec 24, 2011: libarchive 3.0.2 released
Dec 23, 2011: Various fixes merged from FreeBSD
Dec 23, 2011: Symlink support in Zip reader and writer
Dec 23, 2011: Robustness fixes to 7Zip reader

Nov 27, 2011: libarchive 3.0.1b released

Nov 26, 2011: 7Zip reader
Nov 26, 2011: Small fixes to ISO and Zip to improve robustness with corrupted input
Nov 24, 2011: Improve streaming Zip reader's support for uncompressed entries
Nov 20, 2011: New seeking Zip reader supports SFX Zip archives
Nov 20, 2011: Build fixes on Windows

Nov 13, 2011: libarchive 3.0.0a released

Nov 06, 2011: Update shared-library version calculations for libarchive 3.x
Sep 04, 2011: Fix tar -s; follow GNU tar for controlling hardlink/symlink substitutions
Aug 18, 2011: Fix reading ISO images built by NetBSD's mkisofs
Aug 15, 2011: Old archive_read_support_compression_XXX functions are deprecated and
   will disappear in libarchive 4.0.
Jun 26, 2011: RAR reader

stand/libsa/lseek.c
/*	$NetBSD: lseek.c,v 1.4 1997/01/22 00:38:10 cgd Exp $	*/

/*-
 * Copyright (c) 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * The Mach Operating System project at Carnegie-Mellon University.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *
 * Copyright (c) 1989, 1990, 1991 Carnegie Mellon University
 * All Rights Reserved.
 *
 * Author: Alessandro Forin
 *
 * Permission to use, copy, modify and distribute this software and its
 * documentation is hereby granted, provided that both the copyright
 * notice and this permission notice appear in all copies of the
 * software, derivative works or modified versions, and any portions
 * thereof, and that both notices appear in supporting documentation.
 *
 * CARNEGIE MELLON ALLOWS FREE USE OF THIS SOFTWARE IN ITS "AS IS"
 * CONDITION.  CARNEGIE MELLON DISCLAIMS ANY LIABILITY OF ANY KIND FOR
 * ANY DAMAGES WHATSOEVER RESULTING FROM THE USE OF THIS SOFTWARE.
 *
 * Carnegie Mellon requests users of this software to return to
 *
 *  Software Distribution Coordinator  or  Software.Distribution@CS.CMU.EDU
 *  School of Computer Science
 *  Carnegie Mellon University
 *  Pittsburgh PA 15213-3890
 *
 * any improvements or extensions that they make and grant Carnegie the
 * rights to redistribute these changes.
 */

#include "stand.h"

off_t
lseek(int fd, off_t offset, int where)
{
	off_t bufpos, filepos, target;
	struct open_file *f;

	f = fd2open_file(fd);
	if (f == NULL || f->f_flags == 0) {
		errno = EBADF;
		return (-1);
	}

	if (f->f_flags & F_RAW) {
		/*
		 * On RAW devices, update internal offset.
		 */
		switch (where) {
		case SEEK_SET:
			f->f_offset = offset;
			break;
		case SEEK_CUR:
			f->f_offset += offset;
			break;
		default:
			errno = EOFFSET;
			return (-1);
		}
		return (f->f_offset);
	}

	/*
	 * If there is some unconsumed data in the readahead buffer and it
	 * contains the desired offset, simply adjust the buffer offset and
	 * length.  We don't bother with SEEK_END here, since the code to
	 * handle it would fail in the same cases where the non-readahead
	 * code fails (namely, for streams which cannot seek backward and whose
	 * size isn't known in advance).
	 */
	if (f->f_ralen != 0 && where != SEEK_END) {
		filepos = (f->f_ops->fo_seek)(f, 0, SEEK_CUR);
		if (filepos == -1)
			return (-1);
		bufpos = filepos - f->f_ralen;
		switch (where) {
		case SEEK_SET:
			target = offset;
			break;
		case SEEK_CUR:
			target = bufpos + offset;
			break;
		default:
			errno = EINVAL;
			return (-1);
		}
		if (bufpos <= target && target < filepos) {
			f->f_raoffset += target - bufpos;
			f->f_ralen -= target - bufpos;
			return (target);
		}
	}

	/*
	 * If this is a relative seek, we need to correct the offset for
	 * bytes that we have already read but the caller doesn't know
	 * about.
	 */
	if (where == SEEK_CUR)
		offset -= f->f_ralen;

	/*
	 * Invalidate the readahead buffer.
	 */
	f->f_ralen = 0;

	return (f->f_ops->fo_seek)(f, offset, where);
}
